### DISCUSSION TITLE
java.lang.IllegalStateException: Kafka server is still starting up, cannot shut down!

### DISCUSSION THREAD

**Discussion by fritz-net** [strimzi/strimzi-kafka-operator#5662] [Q&A]
I'm trying to get started with strimzi, however my kafka containers keep crashing. I use the example for 3 replicas without storage.

```
minikube kubectl -- get pods -A
NAMESPACE              NAME                                         READY   STATUS             RESTARTS       AGE
ingress-nginx          ingress-nginx-controller-69bdbc4d57-bg2jf    1/1     Running            0              14d
kafka                  my-cluster-kafka-0                           0/1     CrashLoopBackOff   5 (89s ago)    7m38s
kafka                  my-cluster-kafka-1                           0/1     Running            7 (2m7s ago)   7m38s
kafka                  my-cluster-kafka-2                           0/1     CrashLoopBackOff   6 (20s ago)    7m38s
kafka                  my-cluster-zookeeper-0                       1/1     Running            0              8m21s
kafka                  my-cluster-zookeeper-1                       1/1     Running            0              8m21s
kafka                  my-cluster-zookeeper-2                       1/1     Running            0              8m21s
kafka                  strimzi-cluster-operator-7695cb5f7f-ld9q5    1/1     Running            0              110m
kube-system            coredns-78fcd69978-6c959                     1/1     Running            2 (14d ago)    28d
kube-system            etcd-minikube                                1/1     Running            2 (14d ago)    28d
kube-system            kindnet-8n68w                                1/1     Running            1 (14d ago)    24d
kube-system            kindnet-nrqqh                                1/1     Running            2 (14d ago)    28d
kube-system            kube-apiserver-minikube                      1/1     Running            2 (14d ago)    28d
kube-system            kube-controller-manager-minikube             1/1     Running            2 (14d ago)    28d
kube-system            kube-proxy-mzwlw                             1/1     Running            2 (14d ago)    28d
kube-system            kube-proxy-v6dzs                             1/1     Running            1 (14d ago)    24d
kube-system            kube-scheduler-minikube                      1/1     Running            2 (14d ago)    28d
kube-system            storage-provisioner                          1/1     Running            3 (14d ago)    28d
kubernetes-dashboard   dashboard-metrics-scraper-5594458c94-kszqh   1/1     Running            0              14d
kubernetes-dashboard   kubernetes-dashboard-654cf69797-6xn6h        1/1     Running            0              14d
```

**Describe the bug**
When starting the kafka cluster via strimzi the kafka containers keep crashing with:
`java.lang.IllegalStateException: Kafka server is still starting up, cannot shut down!`

**To Reproduce**
```
minikube kubectl -- create namespace kafka
minikube kubectl -- create -f 'https://strimzi.io/install/latest?namespace=kafka' -n kafka
# download https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/0.25.0/examples/kafka/kafka-ephemeral.yaml
minikube kubectl -- apply -f  ./../kafka/kafka-ephemeral.yaml -n kafka
```

**Expected behavior**
cluster starting up

**Environment (please complete the following information):**
 - Strimzi version: latest
 - Installation method: yaml
 - Kubernetes cluster: Kubernetes
 - Infrastructure: minikube version: v1.23.1, single node, virtualbox, win10

**YAML files and logs**
```
2021-10-05 20:44:07,268 INFO Client environment:java.library.path=/usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,268 INFO Client environment:java.io.tmpdir=/tmp (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,268 INFO Client environment:java.compiler=<NA> (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,268 INFO Client environment:os.name=Linux (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,268 INFO Client environment:os.arch=amd64 (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,268 INFO Client environment:os.version=4.19.202 (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,269 INFO Client environment:user.name=kafka (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,269 INFO Client environment:user.home=/home/kafka (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,269 INFO Client environment:user.dir=/opt/kafka (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,269 INFO Client environment:os.memory.free=200MB (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,269 INFO Client environment:os.memory.max=974MB (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,269 INFO Client environment:os.memory.total=217MB (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:07,309 INFO Initiating client connection, connectString=my-cluster-zookeeper-client:2181 sessionTimeout=18000 watcher=kafka.zookeeper.ZooKeeperClient$ZooKeeperClientWatcher$@1c39680d (org.apache.zookeeper.ZooKeeper) [main]
2021-10-05 20:44:08,163 INFO jute.maxbuffer value is 4194304 Bytes (org.apache.zookeeper.ClientCnxnSocket) [main]
2021-10-05 20:44:08,220 INFO zookeeper.request.timeout value is 0. feature enabled= (org.apache.zookeeper.ClientCnxn) [main]
2021-10-05 20:44:08,242 INFO Starting poller (io.strimzi.kafka.agent.KafkaAgent) [main]
2021-10-05 20:44:08,261 INFO [ZooKeeperClient Kafka server] Waiting until connected. (kafka.zookeeper.ZooKeeperClient) [main]
2021-10-05 20:44:08,316 INFO Opening socket connection to server my-cluster-zookeeper-client/10.102.234.39:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(my-cluster-zookeeper-client:2181)]
2021-10-05 20:44:14,265 INFO Terminating process due to signal SIGTERM (org.apache.kafka.common.utils.LoggingSignalHandler) [SIGTERM handler]
2021-10-05 20:44:14,399 INFO shutting down (kafka.server.KafkaServer) [kafka-shutdown-hook]
2021-10-05 20:44:14,528 ERROR Fatal error during KafkaServer shutdown. (kafka.server.KafkaServer) [kafka-shutdown-hook]
java.lang.IllegalStateException: Kafka server is still starting up, cannot shut down!
        at kafka.server.KafkaServer.shutdown(KafkaServer.scala:660)
        at kafka.Kafka$.$anonfun$main$3(Kafka.scala:100)
        at kafka.utils.Exit$.$anonfun$addShutdownHook$1(Exit.scala:38)
        at java.base/java.lang.Thread.run(Thread.java:829)
2021-10-05 20:44:14,547 ERROR Halting Kafka. (kafka.Kafka$) [kafka-shutdown-hook]

```

the following line also looks strange to me:
`2021-10-05 20:44:08,316 INFO Opening socket connection to server my-cluster-zookeeper-client/10.102.234.39:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(my-cluster-zookeeper-client:2181)]`

**Comment by scholzj** [comment]
Can you share the full log from the start? Sharing just a a small snippet makes it quite hard to see what is really happening.

**Comment by fritz-net** [comment]
Thank you for your fast response. When pasting only parts of the log I thought I got everything important - sorry for that :(

I let everything running over night, now I don't find the given line anymore. The Kafka logs show no errors (or I don't see any), however they keep restarting. In Zookeeper I see connection timeouts.

I made a script to collect the logs (which I attached).
[my-cluster-kafka-0.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292338/my-cluster-kafka-0.txt)
[my-cluster-kafka-1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292339/my-cluster-kafka-1.txt)
[my-cluster-kafka-2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292340/my-cluster-kafka-2.txt)
[my-cluster-zookeeper-0.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292341/my-cluster-zookeeper-0.txt)
[my-cluster-zookeeper-1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292342/my-cluster-zookeeper-1.txt)
[my-cluster-zookeeper-2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292343/my-cluster-zookeeper-2.txt)
[pods.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292344/pods.txt)

```
#!/bin/bash

rm -Rf ./strimzi/
mkdir -p ./strimzi/

minikube kubectl -- get pods -n kafka > ./strimzi/pods.txt
minikube kubectl -- logs my-cluster-kafka-0 -n kafka > ./strimzi/my-cluster-kafka-0.txt
minikube kubectl -- logs my-cluster-kafka-1 -n kafka > ./strimzi/my-cluster-kafka-1.txt
minikube kubectl -- logs my-cluster-kafka-2 -n kafka > ./strimzi/my-cluster-kafka-2.txt
minikube kubectl -- logs my-cluster-zookeeper-0 -n kafka > ./strimzi/my-cluster-zookeeper-0.txt
minikube kubectl -- logs my-cluster-zookeeper-1 -n kafka > ./strimzi/my-cluster-zookeeper-1.txt
minikube kubectl -- logs my-cluster-zookeeper-2 -n kafka > ./strimzi/my-cluster-zookeeper-2.txt
```


This is the yaml I used - should be the example
```
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    version: 2.8.0
    replicas: 3
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
      - name: tls
        port: 9093
        type: internal
        tls: true
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      log.message.format.version: "2.8"
      inter.broker.protocol.version: "2.8"
    storage:
      type: ephemeral
  zookeeper:
    replicas: 3
    storage:
      type: ephemeral
  entityOperator:
    topicOperator: {}
    userOperator: {}
```

**Comment by fritz-net** [comment]
I freshly applied kafka again to create the requested logs:
[my-cluster-kafka-0.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292465/my-cluster-kafka-0.txt)
[my-cluster-kafka-1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292466/my-cluster-kafka-1.txt)
[my-cluster-kafka-2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292467/my-cluster-kafka-2.txt)
[my-cluster-zookeeper-0.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292468/my-cluster-zookeeper-0.txt)
[my-cluster-zookeeper-1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292469/my-cluster-zookeeper-1.txt)
[my-cluster-zookeeper-2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292470/my-cluster-zookeeper-2.txt)
[pods.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7292471/pods.txt)

All 3 kafka containers show the same error


EDIT: I hope I did not waste your time, but could this be a Memory issue? I would expect an OutOfMemory error or something in this case

**Reply by scholzj** [reply]
I checked the logs, but there does not seem to be anything obvious there. It just seems to be _too slow_ to start. That might come to resources - both memory and CPU. It might not necessarily run out of memory, just be slow garbage collecting etc. How much resources does your Minikube have? Kafka can be fairly hungry. I normally run it for development with 1Gi RAM/0.5 CPU for ZooKeeper node and 2Gi RAM and 0.5-1 CPU for Kafka broker. 

You can also try to configure the healtchecks to give it more time to start, since it seems to be the healthcheck killing it I guess, but not sure ho wmuch it helps: https://strimzi.io/docs/operators/latest/full/using.html#con-common-configuration-healthchecks-reference

**Reply by guimou** [reply]
Hi! Just to let you know @scholzj , I just had the exact same issue on an OpenShift cluster with 50 CPUs and 200GB available (brand new RHPDS instance), so plenty of room, with the default resources for Kafka/Zookeeper.
I tried restarting the pods, even flushed the StatefulSet to let the operator recreate it, same issue. Solved it with a custom readiness probe with an initial delay set at 30s.
So it maybe that it some cases (which?, I don't know exactly though) the default delay (15s iirc) is a little bit tight.
Not sure if a bug needs to be reopened, but I wanted to leave this trace.

**Reply by scholzj** [reply]
Sorry, I don't follow on the relationship to this - I probably don't remember enough details of the this discussion. In general:
* It doesn't matter how many resources you have in your cluster but how many you give to the Kafka pods
* Yes, different cluster might need different probe configurations -> it depends on many aspects, but at the end this is why you can configure the delays and timeouts etc in the Kafka custom resource

**Reply by guimou** [reply]
No worries :-), I saw it was an old conversation.
I just mentioned my overall resources because you asked about the Minikube resources.
And yes, different clusters might need different probe configs. I just wanted to leave this if others like me stumble on this discussion after facing the problem. That said, you may want to keep this in mind as the "default" readiness probe config may fail on normally pretty standard OCP clusters like the one at the base of RH's demo systems. Worth considering jacking it up by a few seconds if this occurs again/often, but that's your call.

**Reply by Keegan-Bantom** [reply]
I had the same issue, increasing the liveness probe "initialDelaySeconds: 120" and readiness probe to "initialDelaySeconds: 120" did the trick for me.

**Reply by TheDevarshiShah** [reply]
Hello, I got the same issue. I've 16 Kafka pods, each having 5 CPU (limit=10) and 15 Gi Memory (limit=30 Gi). Do you think resources are adequate? (All the nodes on which these pods are running, are not having resource crunch.)
